app/settings/surveys/survey-editor.directive.js

```diff
@@ -552,94 +552,93 @@ function SurveyEditorController(
     //Start - modify Survey
 
     function saveSurvey() {
+        // Set saving to true to disable user actions
         $scope.saving_survey = true;
-        // Saving a survey is a 3 step process
-        // First save the actual survey
+        // Save the survey
         FormEndpoint
         .saveCache($scope.survey)
         .$promise
         .then(function (survey) {
-            $scope.survey.id = survey.id;
-            // Second save the survey tasks
-            saveTasks();
-            saveRoles();
-        }, handleResponseErrors);
-        $scope.saving_survey = false;
+            // If the survey is new, cache the new id
+            if ($scope.survey.id !== survey.id) {
+                $scope.survey.id = survey.id;
+            }
+            // Save tasks and return promises
+            return saveTasks();
```

Comment:
You should still be able to run these in parallel, using `return $q.all([saveTasks(), saveRoles()]);` ... maybe including saveAttributes too.. I'm not sure if that relies on tasks saving first?